yarn-project/noir-contracts/src/contracts/token_bridge_contract/src/main.nr

```diff
@@ -0,0 +1,169 @@
+    ) -> Field {
+        let storage = Storage::init(Context::public(&mut context));
+
+        let content_hash = get_mint_content_hash(amount, context.msg_sender(), canceller.address);
+        // Consume message and emit nullifier
+        context.consume_l1_to_l2_message(msg_key, content_hash, secret);
+
+        // Mint tokens 
+        Token::at(storage.token.read()).mint_public(context, context.msg_sender(), amount);
+
+        1
+    }
+
+    // Burns the appropriate amount of tokens and creates a L2 to L1 withdraw message publicly
+    // Requires `from` to give approval to the bridge to burn tokens on their behalf using witness signatures 
+    #[aztec(public)]
+    fn withdraw_public(
+        amount: Field,
+        recipient: EthereumAddress, // ethereum address to withdraw to
+        callerOnL1: EthereumAddress, // ethereum address that can call this function on the L1 portal (0x0 if anyone can call)
+        nonce: Field,
+    ) -> Field {
+        let storage = Storage::init(Context::public(&mut context));
+
+        // Burn tokens 
+        Token::at(storage.token.read()).burn_public(context, context.msg_sender(), amount, nonce);
+
+        // Send an L2 to L1 message
+        let content = get_withdraw_content_hash(amount, recipient.address, callerOnL1.address);
+        context.message_portal(content);        
+        1
+    }
+
+    // Consumes a L1->L2 message and calls the token contract to mint the appropriate amount in private assets
+    // User needs to call token.redeem_shield() to get the private assets
+    #[aztec(private)]
+    fn deposit_private(
+        amount: Field,
+        msg_key: Field, // L1 to L2 message key as derived from the inbox contract
+        secret: Field,
+        canceller: EthereumAddress,
+    ) -> Field {
+        // Consume L1 to L2 message and emit nullifier
+        let content_hash = get_mint_content_hash(amount, context.msg_sender(), canceller.address);
+        context.consume_l1_to_l2_message(inputs, msg_key, content_hash, secret);
+
+        // Mint tokens on L2 
+        // We hash the secret privately and send the hash in a public call so the secret isn't leaked
+        // Furthermore, `mint_private` on token is public. So we can an internal public function 
+        // which then calls the token contract
+        let secret_hash = compute_secret_hash(secret);
+        context.call_public_function(
+            context.this_address(),
+            compute_selector("_call_mint_on_token(Field,Field)"),
+            [amount, secret_hash],
+        );
+
+        1
+    }
+
+    // Burns the appropriate amount of tokens and creates a L2 to L1 withdraw message privately
+    // Requires `from` to give approval to the bridge to burn tokens on their behalf using witness signatures 
+    #[aztec(private)]
+    fn withdraw_private(
+        token: AztecAddress,
+        amount: Field,
+        recipient: EthereumAddress, // ethereum address to withdraw to
```

Comment:
Good use of the types, makes it easier to spot if things make sense instead of Fields all over the place.